src/govuk/components/components.template.test.js

```diff
@@ -0,0 +1,124 @@
+        rules: {
+          // We don't use boolean attributes consistently – buttons currently
+          // use disabled="disabled"
+          'attribute-boolean-style': 'off',
+
+          // Allow pattern attribute on input type="number"
+          'input-attributes': 'off',
+
+          // Allow for conditional comments (used in header for fallback png)
+          'no-conditional-comment': 'off',
+
+          // Allow inline styles for testing purposes
+          'no-inline-style': 'off',
+
+          // Allow for explicit roles on regions that have implict roles
+          // We do this to better support AT with older versions of IE that
+          // have partial support for HTML5 semantic elements
+          'no-redundant-role': 'off',
+
+          // More hassle than it's worth 👾
+          'no-trailing-whitespace': 'off',
+
+          // We still support creating `input type=button` with the button
+          // component, but you have to explicitly choose to use them over
+          // buttons
+          'prefer-button': 'off',
+
+          // Allow use of roles where there are native elements that would give
+          // us that role automatically, e.g. <section> instead of
+          // <div role="region">
+          //
+          // This is mainly needed for links styled as buttons, but we do this
+          // in the cookie banner and notification banner too
+          'prefer-native-element': 'off',
+
+          // HTML Validate is opinionated about IDs beginning with a letter and
+          // only containing letters, numbers, underscores and dashes – which is
+          // more restrictive than the spec allows.
+          //
+          // Relax the rule to allow anything that is valid according to the
+          // spec.
+          'valid-id': ['error', { relaxed: true }]
+        },
+        elements: [
+          'html5',
+          {
+            // Allow textarea autocomplete attribute to be street-address
+            // (html-validate only allows on/off in default rules)
+            textarea: {
+              attributes: {
+                autocomplete: { enum: ['on', 'off', 'street-address'] }
+              }
+            },
+            // Allow buttons to omit the type attribute (defaults to 'submit')
+            button: {
+              attributes: {
+                type: { required: false }
+              }
+            }
+          }
+        ]
+      })
+    })
+
+    it('renders valid HTML for each component example', () => {
+      const componentTasks = allComponents.map(async (component) => {
+        const { examples } = await getComponentData(component)
+
+        // Loop through component examples
+        examples.forEach(({ name, data }) => {
+          const html = renderHtml(component, data)
+          const report = validator.validateString(html)
+
+          expect({ component, name, report })
+            .toEqual({ component, name, report: expect.objectContaining({ valid: true }) })
```

Comment:
As flagged on Slack, this doesn't appear to be working as intended – if I intentionally make a component output invalid HTML the tests are still passing.

Reply:
Thanks for spotting this one! A bit of digging revealed `new HtmlValidate()` can't just take a "config override", it also needs the default `extends: ['html-validate:recommended']` preset too which describes the test better

Fixed now 🙈 

It also looks like you can pass the **HtmlValidate** instance into `.toHTMLValidate()` so I might go back to doing that, as this will keep the nicely formatted Jest output

Reply:
Ace, that sounds like a good improvement 👍🏻

Reply:
Ahh I'm afraid it doesn't support that. It's possible to pass in a config, but not a validator

Are you happy leaving it as a diff of the whole report object? It's definitely nerdier, and you get the full HTML source showing up in the results array (quite useful).

### Before
```
Result should be valid but had error "Unexpected close-tag, expected opening tag."
```

### After
<img width="1188" alt="Image showing HTML validator report object diff" src="https://user-images.githubusercontent.com/415517/193247689-7d5bb913-0662-43a4-b9d4-aa661efebc48.png">